README.md
# homebridge-macosx-info
[![npm](https://img.shields.io/npm/dt/homebridge-macosx-info.svg)](https://www.npmjs.com/package/homebridge-macosx-info) 
[![npm](https://img.shields.io/npm/v/homebridge-macosx-info.svg)](https://www.npmjs.com/package/homebridge-macosx-info)
[![GitHub license](https://img.shields.io/github/license/ad5030/homebridge-macosx-info.svg)](https://github.com/ad5030/homebridge-macosx-info)
<!-- [![Donate](https://img.shields.io/badge/donate-paypal-yellowgreen.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=9MC83TRGACQPJ&source=url) -->


*See [changelog](docs/CHANGELOG.md)*

homebridge-macosx-info is homebridge plugin for Apple HomeKit, get and return somes systems informations from macOSX computer. 

Such as :
* updateTime
* Temperature (C°)
* Fan speed (rpm)
* Uptime
  * show how long system has been running
* Load average (%) 
  * the load average of the system over the last 1, 5, and 15 minute
* Free Mem (Mo)
* Disk avalable (%)
* Users (nb)
* CPU Power consumption (Watt)

You can see below screenshots for illustrate homebridge-macos-info plugin.

![homebridge-macos-info, Eve., screenshot](/img/homebridge-macosx-info_screenshots.png)

>Screenshots are taken from the Elgato Eve.app

## Exemple of .json data response file
```json  
{
    "updateTime":"Fri May 31 19:35:36 CEST 2019",
    "temperature":31.3,
    "fan":1797,
    "power":1.25,
    "uptime":"up 13:21",
    "load":"2.52 2.17 2.07",
    "freemem":639.96,
    "disk":"50",
    "user":2
}
```
## Prerequisites
* Install [Homebrew](https://brew.sh)<span style="color:gray"> *(Homebrew installs the stuff you need that Apple didn’t)*</span>
* Install [node.js](https://nodejs.org/en/download/package-manager/#macos) on macOS
* Install [Homebridge](https://github.com/nfarina/homebridge/wiki/Install-Homebridge-on-macOS) on macOS
* Install [Homebridge Config UI X](https://github.com/oznu/homebridge-config-ui-x#readme) on macOS <span style="color:gray">*(optional)</span>*
* Install [Eve.app](https://www.evehome.com/en/eve-app) on iOS (for all availables plugin function), or it's possible to used "Home" app, but only on macOSX Majave and iOS (all plugin function aren't availables on this app !)
* [Enable NOPASSWD](#STEP-3-:-Add-NOPASSWD-entry-in-/etc/sudoers) for user in `/etc/sudoers` file

## Installation
Used [npm](https://www.npmjs.com/package/homebridge-macosx-info) tool to install homebridge-macosx-info, and execute the command line below

```console
npm i homebridge-macosx-info
```

## Configuration
### STEP 1 : homebridge config.json file
Add this lines in config.json
```json    
"accessories": [
        {
            "accessory": "MacOSXSysInfo",
            "name": "macOSX Info",
            "file": "/tmp/_homebridge-macosx-info.json",
            "serial": "042-03-000",
            "consumption": true,
            "user": true,
            "updateInterval": 60000
        }
    ],
```

| Parameter       | Note | Optionnal | value | 
|-----------------|------|-----------|-------|
| `accessory`     | Name of accessory|No|`MacOSXSysInfo`|
| `name`          | a human-readable name for your plugin|No|`macOSX Info`|
| `file`          | .json respons file|yes|default : `/tmp/_homebridge-macosx-info.json`|
| `updateInterval`| is time in ms of data update|yes|default : `null`|
| `consumption`| `true` for log CPU Consumption|yes|default : `null`|
| `user`| `true` for log Users number|yes|default : `null`|


>Note : 
>1. The `index.js` call *`<PATH of Node Module>/homebridge-macosx-info/sh/homebridge-macosx-info.sh`* shell script. You can find this script in the repository in `/src/sh` directory
>2. It's possible that you can change the path of `homebridge-macosx-info.sh` in `readUptime` function on `index.js` script
```js
async function readUptime() {
    const exec = require('child_process').exec;
    var script = await exec('/usr/local/lib/node_modules/homebridge-macosx-info/src/sh/homebridge-macosx-info.sh',
        (error, stdout, stderr) => {
            if (error !== null) {
                //this.log("exec error: " + ${error});
            }
        }); 
};
```
### STEP 2 : Adapte "homebridge-macosx-info.sh" file in "src/sh" directory
1. You can change path of temporary .json files -> var `JSON_DATA_FILE`
2. You can change path and option of [`check_osx_smc`](https://github.com/jedda/OSX-Monitoring-Tools/tree/master/check_osx_smc) binary -> var `CHECK_OSX_SMC`

```sh
DIR=$(dirname $0)
JSON_DATA_FILE=/tmp/_homebridge-macosx-info.json # path of .json respons file 
CHECK_OSX_SMC=$DIR/../../bin/check_osx_smc # path of check_osx_smc binary

function sys_mon()
{
    # See the hardware compatibility -> https://github.com/jedda/OSX-Monitoring-Tools/blob/master/check_osx_smc/known-registers.md
    # See README -> https://github.com/jedda/OSX-Monitoring-Tools/blob/master/check_osx_smc/README.md
    read -a fields <<< `$CHECK_OSX_SMC -s c -r TA0P,F0Ac -w 70,5200 -c 85,5800`
    _temp=${fields[7]//,/.}
    _fan=${fields[8]}

    _time=`date`
    read -a fields <<< `sudo powermetrics -i 500 -n1 --samplers cpu_power | grep "CPUs+GT+SA" | sed 's/Intel energy model derived package power (CPUs+GT+SA): //g'`
    _power=${fields[0]//W/}

    _uptime=`uptime`
    _load=$_uptime

    _uptime=${_uptime%users*} ; _uptime=${_uptime%,*} ; _uptime=${_uptime#*up} ; _uptime=${_uptime%,*} ; _uptime=${_uptime#*up} ; _uptime="up ${_uptime# }"
    _load=${_load#*load averages: }

    _user=`who | wc -l`
    _user="${_user// /}"

    read -a fields <<< `vm_stat | perl -ne '/page size of (\d+)/ and $size=$1; /Pages\s+([^:]+)[^\d]+(\d+)/ and printf("%-16s % 16.2f Mi\n", "$1:", $2 * $size / 1048576)' | grep "free:"` ; _freemem=${fields[1]}
    read -a fields <<<  `df -h / | grep /` ; _disk=${fields[4]//%/}

    echo '{"updateTime":"'${_time}'","temperature":'${_temp:5:4}',"fan":'${_fan:5:4}',"power":'${_power}',"uptime":"'${_uptime}'","load":"'${_load}'","freemem":'${_freemem:0:6}',"disk":"'${_disk}'","user":'${_user}'}' > $JSON_DATA_FILE
}
```
### STEP 3 : Add NOPASSWD entry in your /etc/sudoers 
```sh
# root and users in group wheel can run anything on any machine as any user
root        ALL = (ALL) ALL
%admin      ALL = (ALL) ALL
<USER>      ALL=NOPASSWD: ALL
```
>**Note :** 
You must change the user `<USER>` by the user who run `homebridge` in your system

### STEP 4 : restart homebridge 
Combine the two commands in a terminal to restart homebridge background process

 - `launchctl unload ~/Library/LaunchAgents/com.homebridge.server.plist`
 - `launchctl load ~/Library/LaunchAgents/com.homebridge.server.plist`

>**Note :** 
This commands are only avalable for macOS 

## Todo
- [x] Generate all the measures in a .json file [[#3]](https://github.com/ad5030/homebridge-macosx-info/issues/3)
- [x] Worked on performance
  - [x] Use only sh built-in (no sed & no awk) [[#4]](https://github.com/ad5030/homebridge-macosx-info/issues/3)

## Known bugs
- [x] Uptime error in "homebridge-macosx-info" after more than one day ! [[#1]](https://github.com/ad5030/homebridge-macosx-info/issues/1)
- [x] Temparature and fan mesures don't work on all Apple mac hardware. Used now [`check_osx_smc`](https://github.com/jedda/OSX-Monitoring-Tools/tree/master/check_osx_smc) binary. You can see the hardware compatibility [here](https://github.com/jedda/OSX-Monitoring-Tools/blob/master/check_osx_smc/known-registers.md) [[#2]](https://github.com/ad5030/homebridge-macosx-info/issues/2)

## Credits
* The original HomeKit API work was done by [KhaosT](https://twitter.com/khaost) in his [HAP-NodeJS](https://github.com/KhaosT/HAP-NodeJS) project
* [simont77 - fakegato-history](https://github.com/simont77/fakegato-history)
* [Jedda Wignall - OSX-Monitoring-Tools/check_osx_smc](https://github.com/jedda/OSX-Monitoring-Tools/tree/master/check_osx_smc)

## Disclaimer
I'm furnishing this software "as is". I do not provide any warranty of the item whatsoever, whether express, implied, or statutory, including, but not limited to, any warranty of merchantability or fitness for a particular purpose or any warranty that the contents of the item will be error-free. The development of this module is not supported by Apple Inc. or eve. These vendors and me are not responsible for direct, indirect, incidental or consequential damages resulting from any defect, error or failure to perform.

## License
This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details

docs/CHANGELOG.md
# homebridge-macosx-info Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## v1.0.3 2019-09-29
- Minor change in package.json

## v1.0.2 2019-08-14
- Minor change in README and code

## v1.0.1 2019-06-07
- Add check_osx_smc binary in git package
- Add relative path for cal check_osx_smc 

## v1.0.0 2019-05-31 
- First released
- This update includes a number of bug fixes and stability improvements

## v0.3.1 2019-05-25 
- Performance : Optimize `homebridge-macosx-info.sh` script
- Now, CPU Power consumption and Users are optional. The confuguration is in `config.json` homebridge file

## v0.3.0 2019-05-18 
- Add CPU Power consumption
- Add units constant in index.js
- Add Prerequisites in README.md

## v0.2.11 2019-05-16 
- Added nunber of Users connected in the host
- fix bug in uptime return 
  
## v0.2.10 2019-05-10 
- Deleted `serial` entry in `config.json`
- Minor change in `README.md` to specify installation steps
  
## v0.2.9 2019-05-09 
- Minor change in `README.md` to specify installation steps
- Add constants `JSON_DATA_FILE` & `CHECK_OSX_SMC` in the shell script

## v0.2.8 2019-05-08
- Minor change in `index.js` to call `homebridge-macosx-info.sh`
- Minor change in `.json` respons file
- Performance
  - Now no `sed` & `awk` in shell script
- Added the compatibility list of Apple mac hardware (with check_osx_smc)
- Added CHANGELOG.md file
- Complete the README.md file
- Opitimized screenshots

## v0.2.6 2019-05-05
 - Firt version in github public repository 



private/test.js
var Accessory, Service, Characteristic, UUIDGen, FakeGatoHistoryService;
var inherits = require('util').inherits;
const fs = require('fs');
const packageFile = require("./package.json");
var os = require("os");
var hostname = os.hostname();

module.exports = function(homebridge) {
    if(!isConfig(homebridge.user.configPath(), "accessories", "MacOSXSysInfo")) {
        return;
    }
    
    Accessory = homebridge.platformAccessory;
    Service = homebridge.hap.Service;
    Characteristic = homebridge.hap.Characteristic;
    UUIDGen = homebridge.hap.uuid;
	FakeGatoHistoryService = require("fakegato-history")(homebridge);

    homebridge.registerAccessory('homebridge-macosx-info', 'MacOSXSysInfo', MacOSXSysInfo);
}

function readUptime() {
	const exec = require('child_process').exec;
	//var _cmd=this.cmd;  
	//var script = exec('_cmd',
	var script = exec('~/r2d2/it/homekit/homebridge-macosx-info/sh/homebridge-macosx-info.sh',
		(error, stdout, stderr) => {
			if (error !== null) {
				//this.log("exec error: " + ${error});
			}
		});			
};

function isConfig(configFile, type, name) {
    var config = JSON.parse(fs.readFileSync(configFile));
    if("accessories" === type) {
        var accessories = config.accessories;
        for(var i in accessories) {
            if(accessories[i]['accessory'] === name) {
                return true;
            }
        }
    } else if("platforms" === type) {
        var platforms = config.platforms;
        for(var i in platforms) {
            if(platforms[i]['platform'] === name) {
                return true;
            }
        }
    } else {
    }
    return false;
};

function MacOSXSysInfo(log, config) {
    if(null == config) {
        return;
    }

    this.log = log;
	this.name = config["name"];
	this.serial = config["serial"];
	//this.cmd = config["cmd"];
    if(config["file"]) {
        this.readFile = config["file"];
    } else {
        this.readFile = "/tmp/homebridge-macosx-info.json";
	}
  if(config["updateInterval"] && config["updateInterval"] > 0) {
        this.updateInterval = config["updateInterval"];
    } else {
        this.updateInterval = null;
    }
  
	this.setUpServices();
};

//MacOSXSysInfo.prototype.getUptime = function (callback) {
//	
//	var data = fs.readFileSync("/tmp/_hb_uptime.txt", "utf-8");
//	var uptime = data.substring(10, data.indexOf(",", data.indexOf(",", 0)+1));
		
//	callback(null, uptime);
//};

MacOSXSysInfo.prototype.getUptime = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var uptime = (obj.uptime);
	callback(null, uptime);
};

MacOSXSysInfo.prototype.getFan = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var fan = parseFloat(obj.fan);
	callback(null, fan);
};

MacOSXSysInfo.prototype.getDisk = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var disk = parseFloat(obj.disk);
	callback(null, disk);
};

MacOSXSysInfo.prototype.getAvgLoad = function (callback) {
//	var data = fs.readFileSync("/tmp/_hb_uptime.txt", "utf-8");
//	var load = data.substring(data.length - 15);
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var load = obj.load;
	
	callback(null, load);
};

//MacOSXSysInfo.prototype.getAvgLoad = function (callback) {
	
//	var json = fs.readFileSync("/tmp/sys_mon.json", "utf-8");
	//var json = '{"temperature":28.7, "fan":1796, "uptime":12.5, "load":6.9, "mem":1163}';
//	var obj = JSON.parse(json);
//	var load = parseFloat(obj.load).toFixed(1);
//	callback(null, load);

//};

MacOSXSysInfo.prototype.getMem = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var mem = parseFloat(obj.mem);
	callback(null, mem);

};

MacOSXSysInfo.prototype.setUpServices = function () {

	var that = this;
	var temp;
	
	this.infoService = new Service.AccessoryInformation();
	this.infoService
		.setCharacteristic(Characteristic.Manufacturer, "di-marco_a.net")
		.setCharacteristic(Characteristic.Model, this.name)
		.setCharacteristic(Characteristic.SerialNumber, this.serial + "_" + packageFile.version)
//		.setCharacteristic(Characteristic.SerialNumber, "042-SN-20190407-" + packageFile.version)
		.setCharacteristic(Characteristic.FirmwareRevision, packageFile.version);
	
	this.fakeGatoHistoryService = new FakeGatoHistoryService("weather", this, { storage: 'fs' });
	
	let uuid1 = UUIDGen.generate(that.name + '-Uptime');
	info = function (displayName, subtype) {
		Characteristic.call(this, 'Uptime :', uuid1);
		this.setProps({
			format: Characteristic.Formats.STRING,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(info, Characteristic);
	info.UUID = uuid1;

	let uuid2 = UUIDGen.generate(that.name + '-AvgLoad');
	load = function () {
		Characteristic.call(this, 'Avg Load (%) :', uuid2);
		this.setProps({
			format: Characteristic.Formats.STRING,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(load, Characteristic);
	load.UUID = uuid2;
	
	let uuid3 = UUIDGen.generate(that.name + '-Mem');
	mem = function () {
		Characteristic.call(this, 'Free Mem (Mo) :', uuid3);
		this.setProps({
			format: Characteristic.Formats.STRING,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(mem, Characteristic);
	mem.UUID = uuid3;

	//let uuid4 = UUIDGen.generate(that.name + '-Fan');
	//fan = function () {
	//	Characteristic.call(this, 'Fan (rpm) :', uuid4);
	//	this.setProps({
	//		format: Characteristic.Formats.STRING,
	//		perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
	//	});
	//	this.value = this.getDefaultValue();
	//};
	//inherits(fan, Characteristic);
	//fan.UUID = uuid4;

	let uuid5 = UUIDGen.generate(that.name + '-Disk');
	disk = function () {
		Characteristic.call(this, 'Disk used (%) :', uuid5);
		this.setProps({
			format: Characteristic.Formats.STRING,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(disk, Characteristic);
	disk.UUID = uuid5;

	this.macOSXService = new Service.TemperatureSensor(that.name);
	var currentTemperatureCharacteristic = this.macOSXService.getCharacteristic(Characteristic.CurrentTemperature);

	this.macOSXService = new Service.FanSensor(that.name);
	var currentFanCharacteristic = this.macOSXService.getCharacteristic(Characteristic.CurrentFan);

	this.macOSXService.getCharacteristic(info)
		.on('get', this.getUptime.bind(this));
	this.macOSXService.getCharacteristic(load)
		.on('get', this.getAvgLoad.bind(this));
	this.macOSXService.getCharacteristic(mem)
		.on('get', this.getMem.bind(this));
	//this.macOSXService.getCharacteristic(fan)
	//	.on('get', this.getFan.bind(this));
		this.macOSXService.getCharacteristic(disk)
		.on('get', this.getDisk.bind(this));
		function getCurrentTemperature() {
			var data = fs.readFileSync(that.readFile, "utf-8");

			var obj = JSON.parse(data);
			var temperatureVal = (obj.temperature);

			//var temperatureVal = parseFloat(data) / 1000;
		//	var temperatureVal = parseFloat(data.replace(',', '.'));
			//var temperatureVal = parseFloat(data);
			temp = temperatureVal;
			that.log.debug("update currentTemperatureCharacteristic value: " + temperatureVal);
			return temperatureVal;
	}
	
		//ADM 
		function getCurrentFan() {
			var data = fs.readFileSync(that.readFile, "utf-8");

			var obj = JSON.parse(data);
			var fanVal = (obj.fan);

			fan = fanVal;
			that.log.debug("update currentFanCharacteristic value: " + fanVal);
			return fanVal;
	}

	readUptime();
	
	currentTemperatureCharacteristic.updateValue(getCurrentTemperature());
	if(that.updateInterval) {
		setInterval(() => {
			currentTemperatureCharacteristic.updateValue(getCurrentTemperature());
			
			that.log("Temperature: " + temp);
			this.fakeGatoHistoryService.addEntry({time: new Date().getTime() / 1000, temp: temp});
			//this.fakeGatoHistoryService.addEntry({time: new Date().getTime(), temp: temp});
			
			readUptime();
			
		}, that.updateInterval);
	}

	currentTemperatureCharacteristic.on('get', (callback) => {
		callback(null, getCurrentTemperature());
	//ADM 
	currentFanCharacteristic.updateValue(getCurrentFan());
	if(that.updateInterval) {
		setInterval(() => {
			currentFanCharacteristic.updateValue(getCurrentFan());
			
			that.log("Fan: " + fan);
			this.fakeGatoHistoryService2.addEntry({time: new Date().getTime() / 1000, fan: fan});
			//this.fakeGatoHistoryService.addEntry({time: new Date().getTime(), temp: temp});
			
			readUptime();
			
		}, that.updateInterval);
	}
	
	currentFanCharacteristic.on('get', (callback) => {
		callback(null, getCurrentFan());
	});
}

MacOSXSysInfo.prototype.getServices = function () {

	return [this.infoService, this.fakeGatoHistoryService, this.macOSXService];
};

private/test.sh
#!/bin/bash
 
 
_LOAD=`sysctl -n vm.loadavg` 
_LOAD="${_LOAD//[\{\}]}"
_LOAD="${_LOAD/ /}"
_LOAD="${_LOAD%?}"

echo --$_LOAD--


index.js
//-------------------------------------------------------------------
//~ @(#) Name : index.js
//~ @(#) Desc : 
//~ @(#) version : 1.0
// Auteur : adm@di-marco.net
// Date : 2019-05-05
//-------------------------------------------------------------------
// Version history
//   v1.O - Initial version
//   test and work on : mac mini (late 2014) & macOSX 10.13.6(High Sierra) 
//-------------------------------------------------------------------
var Accessory, Service, Characteristic, UUIDGen, FakeGatoHistoryService;
var inherits = require('util').inherits;
const fs = require('fs');
const packageFile = require("./package.json");
var os = require("os");
var hostname = os.hostname();

const UNIT_RPM='rpm'
const UNIT_MO='Mo'
const UNIT_PERCENT='%'
const UNIT_WATT='Watt'

module.exports = function(homebridge) {
    if(!isConfig(homebridge.user.configPath(), "accessories", "MacOSXSysInfo")) {
        return;
    }
    
    Accessory = homebridge.platformAccessory;
    Service = homebridge.hap.Service;
    Characteristic = homebridge.hap.Characteristic;
    UUIDGen = homebridge.hap.uuid;
    FakeGatoHistoryService = require("fakegato-history")(homebridge);

    homebridge.registerAccessory('homebridge-macosx-info', 'MacOSXSysInfo', MacOSXSysInfo);
}

async function readUptime() {
    const exec = require('child_process').exec;
var script = await exec('/usr/local/lib/node_modules/homebridge-macosx-info/src/sh/homebridge-macosx-info.sh', 
(error, stdout, stderr) => {
            if (error !== null) {
                //this.log("exec error: " + ${error});
            }
        }); 
};

function isConfig(configFile, type, name) {
    var config = JSON.parse(fs.readFileSync(configFile));
    if("accessories" === type) {
        var accessories = config.accessories;
        for(var i in accessories) {
            if(accessories[i]['accessory'] === name) {
                return true;
            }
        }
    } else if("platforms" === type) {
        var platforms = config.platforms;
        for(var i in platforms) {
            if(platforms[i]['platform'] === name) {
                return true;
            }
        }
    } else {
    }
    return false;
};

function MacOSXSysInfo(log, config) {
    if(null == config) {
        return;
    }

    this.log = log;
	this.name = config["name"];
    if(config["file"]) {
		this.readFile = config["file"];
    } else {
		this.readFile = "/tmp/_homebridge-macosx-info.json";
	}
	if(config["consumption"]) {
		this.consumption = config["consumption"];
    } else {
		this.consumption = null;
	}
	if(config["user"]) {
		this.user = config["user"];
    } else {
		this.user = null;
	}
	  if(config["updateInterval"] && config["updateInterval"] > 0) {
        this.updateInterval = config["updateInterval"];
    } else {
        this.updateInterval = null;
    }
  
	this.setUpServices();
};

MacOSXSysInfo.prototype.getUptime = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var uptime = (obj.uptime);
	callback(null, uptime);
};

MacOSXSysInfo.prototype.getFan = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var fan = obj.fan;
	callback(null, fan);
};

MacOSXSysInfo.prototype.getDisk = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var disk = obj.disk;
	callback(null, disk);
};

MacOSXSysInfo.prototype.getAvgLoad = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var load = obj.load;
	
	callback(null, load);
};

MacOSXSysInfo.prototype.getFreeMem = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var freemem = obj.freemem;
	callback(null, freemem);
};

MacOSXSysInfo.prototype.getUser = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var user = obj.user;
	callback(null, user);
};

MacOSXSysInfo.prototype.getPower = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var power = obj.power;
	callback(null, power);
};

MacOSXSysInfo.prototype.setUpServices = function () {

	var that = this;
	var temp;
	
	this.infoService = new Service.AccessoryInformation();
	this.infoService
		.setCharacteristic(Characteristic.Manufacturer, "@ad5030")
		.setCharacteristic(Characteristic.Model, this.name)
		.setCharacteristic(Characteristic.SerialNumber, "042-SN-20190505" + "_" + packageFile.version)
		.setCharacteristic(Characteristic.FirmwareRevision, packageFile.version);
	
	this.fakeGatoHistoryService = new FakeGatoHistoryService("weather", this, { storage: 'fs' });
	
	let uuid1 = UUIDGen.generate(that.name + '-Uptime');
	info = function (displayName, subtype) {
		Characteristic.call(this, 'Uptime', uuid1);
		this.setProps({
			format: Characteristic.Formats.STRING,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(info, Characteristic);
	info.UUID = uuid1;

	let uuid2 = UUIDGen.generate(that.name + '-AvgLoad');
	load = function () {
		Characteristic.call(this, 'Avg Load', uuid2);
		this.setProps({
			format: Characteristic.Formats.STRING,
			unit: UNIT_PERCENT,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(load, Characteristic);
	load.UUID = uuid2;
	
	let uuid3 = UUIDGen.generate(that.name + '-Mem');
	freemem = function () {
		Characteristic.call(this, 'Free Mem', uuid3);
		this.setProps({
			format: Characteristic.Formats.STRING,
			unit: UNIT_MO,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(freemem, Characteristic);
	freemem.UUID = uuid3;

	let uuid4 = UUIDGen.generate(that.name + '-Fan');
	fan = function () {
		Characteristic.call(this, 'Fan speed', uuid4);
		this.setProps({
			format: Characteristic.Formats.STRING,
			unit: UNIT_RPM,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(fan, Characteristic);
	fan.UUID = uuid4;

	let uuid5 = UUIDGen.generate(that.name + '-Disk');
	disk = function () {
		Characteristic.call(this, 'Disk used', uuid5);
		this.setProps({
			format: Characteristic.Formats.STRING,
			unit: UNIT_PERCENT,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(disk, Characteristic);
	disk.UUID = uuid5;

	let uuid6 = UUIDGen.generate(that.name + '-User');
	user = function () {
		Characteristic.call(this, 'Users (nb)', uuid6);
		this.setProps({
			format: Characteristic.Formats.STRING,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(user, Characteristic);
	user.UUID = uuid6;


	let uuid7 = UUIDGen.generate(that.name + '-Consumption');
	power = function () {
		Characteristic.call(this, 'Consumption', uuid7);
		this.setProps({
			format: Characteristic.Formats.STRING,
			unit: UNIT_WATT,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(power, Characteristic);
	power.UUID = uuid7;

	this.macOSXService = new Service.TemperatureSensor(that.name);
	var currentTemperatureCharacteristic = this.macOSXService.getCharacteristic(Characteristic.CurrentTemperature);		
	this.macOSXService.getCharacteristic(info)
		.on('get', this.getUptime.bind(this));
	this.macOSXService.getCharacteristic(load)
		.on('get', this.getAvgLoad.bind(this));
	this.macOSXService.getCharacteristic(freemem)
		.on('get', this.getFreeMem.bind(this));
	this.macOSXService.getCharacteristic(fan)
		.on('get', this.getFan.bind(this));
	this.macOSXService.getCharacteristic(disk)
		.on('get', this.getDisk.bind(this));
	if(this.user) {
	this.macOSXService.getCharacteristic(user)
		.on('get', this.getUser.bind(this));
	}
	if(this.consumption) {
	this.macOSXService.getCharacteristic(power)
		.on('get', this.getPower.bind(this));
	}
	function getCurrentTemperature() {
		var data = fs.readFileSync(that.readFile, "utf-8");

		var obj = JSON.parse(data);
		var temperatureVal = (obj.temperature);

		temp = temperatureVal;
		that.log.debug("update currentTemperatureCharacteristic value: " + temperatureVal);
		return temperatureVal;
	}
	
	readUptime();
	
	currentTemperatureCharacteristic.updateValue(getCurrentTemperature());
	if(that.updateInterval) {
		setInterval(() => {
			currentTemperatureCharacteristic.updateValue(getCurrentTemperature());
			
			that.log("Current Temp: " + temp);
			this.fakeGatoHistoryService.addEntry({time: new Date().getTime() / 1000, temp: temp});
			//this.fakeGatoHistoryService.addEntry({time: new Date().getTime(), temp: temp});	

			readUptime();
			
		}, that.updateInterval);
	}
	
	currentTemperatureCharacteristic.on('get', (callback) => {
		callback(null, getCurrentTemperature());
	});
}

MacOSXSysInfo.prototype.getServices = function () {

	return [this.infoService, this.fakeGatoHistoryService, this.macOSXService];
};